lib/charms/traefik_k8s/v1/ingress.py

```diff
@@ -282,32 +313,29 @@ def is_ready(self, relation: Relation = None):
             log.warning("Requirer not ready; validation error encountered: %s" % str(e))
             return False
 
-    def _provider_app_data(self, relation: Relation) -> ProviderApplicationData:
-        """Fetch and validate this provider's app databag."""
-        if not relation.app or not relation.app.name:
+    def _provided_url(self, relation: Relation) -> ProviderIngressData:
+        """Fetch and validate this app databag; return the ingress url."""
+        if not relation.app or not relation.app.name or not not self.unit.is_leader():
```

Comment:
Please refactor this. `if not ... or not ... or not not ...` is not an intuitive read 😝

Reply:
check!